### PR TITLE
re #2102: Added a #RENAME function, or query parameter rename.fields

### PR DESCRIPTION
**Pull request by ivakegg**
* This will rename fields in the return set

### CODE REVIEW

**Review by apmoriarty — DISMISSED** [review]

**Review by keith-ratcliffe — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/transformer/FieldRenameTransform.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        this.reducedResponse = reducedResponse;
+    }
+
+    @Nullable
+    @Override
+    public Map.Entry<Key,Document> apply(@Nullable Map.Entry<Key,Document> keyDocumentEntry) {
+        if (keyDocumentEntry != null) {
+            Document document = keyDocumentEntry.getValue();
+
+            for (Map.Entry<String,String> rename : this.renameFieldMap.entrySet()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/transformer/FieldRenameTransform.java]: Need to handle grouping context

File: warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java [hunk 2]
```diff
@@ -648,6 +649,17 @@ private void addConfigBasedTransformers() {
                                     .addTransform(new GroupingTransform(groupFields, this.markingFunctions, this.getQueryExecutionForPageTimeout()));
                 }
             }
+
+            if (getConfig().getRenameFields() != null && !getConfig().getRenameFields().isEmpty()) {
+                DocumentTransform alreadyExists = ((DocumentTransformer) this.transformerInstance).containsTransform(FieldRenameTransform.class);
+                if (alreadyExists != null) {
+                    ((FieldRenameTransform) alreadyExists).updateConfig(getConfig().getRenameFields());
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/tables/ShardQueryLogic.java:657]: Lets incorporate the query model mappings in here.  Note that currently the rename will only work on the datawave field names.

**Review by avgAGB — DISMISSED** [review]

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java [hunk 3]
```diff
@@ -34,6 +34,16 @@ public void testOptionsFunction() throws ParseException {
         assertOption("limit.fields", "FOO_1_BAR=3,FOO_1=2");
     }
 
+    @Test
+    public void testRenameFunction() throws ParseException {
+        // Verify that an empty groupby functions results in an empty parameter value.
```
Comment (apmoriarty) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java]: groupby, you say?

**Review by apmoriarty — DISMISSED** [review]
great tests

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java (hunk 3 above)
Comment (ivakegg) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/QueryOptionsFromQueryVisitorTest.java]: updated

**Review by apmoriarty — APPROVED** [review]

**Review by keith-ratcliffe — APPROVED** [review]

**Review by hlgp — APPROVED** [review]

**Pull request merged**